CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,326 @@
+		toKill = true;
+	    
+	    
+	    if(toKill)
+	      {
+		discarded->push_back(pf);
+		oldToNew[i] = (-discarded->size());
+		badToOld.push_back(i);
+		continue;
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	//deal with HF
+	else if(absEta >= 3.)
+	  {
+	    const math::XYZPointF& ecalPoint = pf.positionAtECALEntrance();
+	    GlobalPoint ecalGPoint(ecalPoint.X(),ecalPoint.Y(),ecalPoint.Z());
+	    HcalDetId closestDetId(hgeom->getClosestCell(ecalGPoint));
+	    
+	    if(closestDetId.subdet() == HcalForward)
+	      {
+		HcalDetId hDetId(closestDetId.subdet(),closestDetId.ieta(),closestDetId.iphi(),1); //depth1
+		
+		//raw*calEnergy() is the same as *calEnergy() - no corrections are done for HF
+		float longE  = pf.rawEcalEnergy() + pf.rawHcalEnergy()/2.;  //depth1
+		float shortE = pf.rawHcalEnergy()/2.;                       //depth2
+		
+		float ecalEnergy = pf.rawEcalEnergy();
+		float hcalEnergy = pf.rawHcalEnergy();
+		float totEnergy = ecalEnergy + hcalEnergy;
+		float totEnergyOrig = totEnergy;
+		
+		bool toKill = false;
+			
+		for(auto const& badIt: badChHF_)
+		  {
+		    if ( hDetId.ieta() == badIt.ieta &&
+			 hDetId.iphi() == badIt.iphi )
+		      {
+			LogDebug("PFCandidateRecalibrator") << "==> orig en (tot,H,E): " 
+							    << pf.energy() << " " << pf.rawHcalEnergy() << " " << pf.rawEcalEnergy();
+			if(badIt.depth == 1) //depth1
+			  {
+			    longE *= badIt.ratio;
+			    ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
+			  }
+			else //depth2
+			  {
+			    shortE *= badIt.ratio;
+			    hcalEnergy = 2*shortE;
+			    if(ecalEnergy > 0)
+			      ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
+			  }
+			//kill candidate if goes below thr
+			if((pf.particleId() == reco::PFCandidate::ParticleType::h_HF && shortE < 1.4) || 
+			   (pf.particleId() == reco::PFCandidate::ParticleType::egamma_HF && longE < 1.4))
+			  toKill = true;
+
+			LogDebug("PFCandidateRecalibrator") << "====> ieta,iphi,depth: " 
+							    << badIt.ieta << " " << badIt.iphi << " " << badIt.depth << " corr: " << badIt.ratio;
+			LogDebug("PFCandidateRecalibrator") << "====> recal en (tot,H,E): " 
+							    << totEnergy << " " << hcalEnergy << " " << ecalEnergy;
+
+		      }
+		  }
+		
+		if(toKill == true)
+		  {
+		    discarded->push_back(pf);
+		    oldToNew[i] = (-discarded->size());
+		    badToOld.push_back(i);
+
+		    LogDebug("PFCandidateRecalibrator") << "==> KILLED ";
+		  }
+		else
+		  {
+		    copy->push_back(pf);
+		    oldToNew[i] = (copy->size());
+		    newToOld.push_back(i);
+		    
+		    copy->back().setHcalEnergy(hcalEnergy, hcalEnergy);
+		    copy->back().setEcalEnergy(ecalEnergy, ecalEnergy);
+		    math::XYZTLorentzVector recalibP4(pf.px()*(totEnergy/totEnergyOrig), pf.py()*(totEnergy/totEnergyOrig), pf.pz()*(totEnergy/totEnergyOrig), totEnergy);
```

Comment:
please precompute this ratio instead of copy-paste of the full operation

Reply:
you are probably right.
one simple solution is to leave the pfCandidate untouched if badChHF_ has 0 size.

HF was supposed to be an easy case, but failed to do the reverse engineering of what happens in PF. if anybody has suggestions, I'm happy to try them..

Reply:
how about scaling the p4.energy() as well instead of replacing it with a corrected raw calo?
I'm guessing that all HF cands have a zero mass. For this case p4New = p4Old*corr should also preserve the mass.

Reply:
will change in a way that, in case the badCH vectors have 0 size, the pfCandidate is untouched

Reply:
also will rescale the p4 momentum as suggested